modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+  }
+  if (counter == n_node)
+    inside = true;
+  return inside;
+}
+
+bool
+MeshCut3DUserObject::findElemNormal(const std::vector<Point> & vertices,
+                                    Point & normal,
+                                    Point & center) const
+{
+  unsigned int n_node = vertices.size();
+
+  center(0) = center(1) = center(2) = 0;
+  normal(0) = normal(1) = normal(2) = 0;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+    center += vertices[i];
+  center /= n_node;
+
+  for (unsigned int i = 0; i < n_node; ++i)
+  {
+    unsigned int iplus1(i < n_node - 1 ? i + 1 : 0);
+    Point ray1 = vertices[i] - center;
+    Point ray2 = vertices[iplus1] - center;
+    normal += ray1.cross(ray2);
+  }
+  normal /= n_node;
+  Xfem::normalizePoint(normal);
+  return true; // to be done
+}
+
+void
+MeshCut3DUserObject::findBoundaryNodes()
+// this is a simple algorithm simply based on the added angle = 360 degrees
+// works fine for planar cutting surface
+// for curved cutting surface, need to re-work this subroutine to make it more general
+{
+  _boundary_map.clear();
+
+  unsigned int n_nodes = _cut_mesh->n_nodes();
+  std::vector<Real> angle(n_nodes, 0);
+  std::vector<unsigned int> node_id(_cut_elem_nnode);
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+    angle[i] = 0;
+
+  const MeshBase::element_iterator elem_begin = _cut_mesh->elements_begin();
+  MeshBase::element_iterator elem_end = _cut_mesh->elements_end();
+  MeshBase::element_iterator elem_it = elem_begin;
+
+  for (; elem_it != elem_end; ++elem_it)
+  {
+    Elem * cut_elem = *elem_it;
+
+    // To match the type definition and data structure in Xfem::angle_rad_3d
+    // C array is used here instead of std::array or std::vector
+    Real vertices[_cut_elem_nnode][LIBMESH_DIM];
+
+    for (unsigned int i = 0; i < _cut_elem_nnode; ++i)
+    {
+      Node * this_node = cut_elem->get_node(i);
+      Point & this_point = *this_node;
+      for (unsigned int j = 0; j < LIBMESH_DIM; ++j)
+        vertices[i][j] = this_point(j);
+      node_id[i] = this_node->id();
+    }
+
+    angle[node_id[0]] += Xfem::angle_rad_3d(&vertices[2][0], &vertices[0][0], &vertices[1][0]);
+    angle[node_id[1]] += Xfem::angle_rad_3d(&vertices[0][0], &vertices[1][0], &vertices[2][0]);
+    angle[node_id[2]] += Xfem::angle_rad_3d(&vertices[1][0], &vertices[2][0], &vertices[0][0]);
+  }
+
+  for (unsigned int i = 0; i < n_nodes; ++i)
+  {
+    if (std::abs(angle[i] - libMesh::pi * 2) > libMesh::pi * 2 * 0.001)
+    {
+      std::shared_ptr<CutPoint> ptr(new CutPoint);
+      _boundary_map[i] = ptr;
+      Node * this_node = _cut_mesh->node_ptr(i);
+      mooseAssert(this_node != nullptr, "Node is NULL");
+      Point & this_point = *this_node;
+
+      mooseAssert(_boundary_map.find(i) != _boundary_map.end(),
+                  "_boundary_map does not have this key") for (unsigned int j = 0; j < LIBMESH_DIM;
+                                                               ++j) _boundary_map.find(i)
```

Comment:
Doing a find and then dereferencing is dangerous. You should probably use either the bracket operator, insert, or emplace().